napari/utils/_injection/_processors.py

```diff
@@ -0,0 +1,225 @@
+import sys
+from concurrent.futures import Future
+from contextlib import nullcontext, suppress
+from functools import partial
+from typing import Any, Callable, Dict, Optional, Set, Type, TypeVar
+
+from ... import layers, types, viewer
+from ...layers._source import layer_source
+from ...utils.misc import ensure_list_of_layer_data_tuple
+from ._accessors import _access_viewer
+
+T = TypeVar("T")
+_NULL = object()
+
+
+def _add_layer_data_tuples_to_viewer(
+    data: Any, return_type=None, viewer=None, source: Optional[dict] = None
+):
+    if viewer is None:
+        viewer = _access_viewer()
+    if viewer and data is not None:
+        data = data if isinstance(data, list) else [data]
+        for datum in ensure_list_of_layer_data_tuple(data):
+            # then try to update a viewer layer with the same name.
+            if len(datum) > 1 and (name := datum[1].get("name")):
+                with suppress(KeyError):
+                    layer = viewer.layers[name]
+                    layer.data = datum[0]
+                    for k, v in datum[1].items():
+                        setattr(layer, k, v)
+                    continue
+            with layer_source(**source) if source else nullcontext():
+                # otherwise create a new layer from the layer data
+                viewer._add_layer_from_data(*datum)
+
+
+def _add_layer_data_to_viewer(
+    data: Any,
+    return_type: Any,
```

Comment:
the `data` could be technically, but it's hard because we don't have a globally applicable "type that could be data for some layer".   It's not even `LayerDataProtocol` because of `Shapes`, I supposed you could theoretically come up with some Union of all the things that will possibly work to create data in at least one Layer type, but that seemed a bit out of scope for a private function :)

the `return_type`, should probably be `Type` (since I use `return_type.__name__`), but really it's a `NewType`, which is a function :/ ... so I guess it could either be a big union of all the NewTypes... anyway, you can see why I punted :joy: